resolwe_bio/processes/expression/expression_aggregator.py

```diff
@@ -0,0 +1,295 @@
+"""Collect expression data from samples grouped by sample annotation field."""
+
+import csv
+import gzip
+import json
+import math
+
+import numpy as np
+
+from resolwe.process import (
+    DataField,
+    FileField,
+    JsonField,
+    ListField,
+    Process,
+    SchedulingClass,
+    StringField,
+)
+
+
+def load_expression(fn=None, sep="\t"):
+    """Read expressions from file."""
+    with gzip.open(fn, "rt") as f:
+        reader = csv.DictReader(f, delimiter=sep)
+        return {row["Gene"]: float(row["Expression"]) for row in reader}
+
+
+def get_indices(descriptors, descriptor):
+    """Return positions of my_element in my_list."""
+    return {i for i, x in enumerate(descriptors) if x == descriptor}
+
+
+def get_values(expressions, descriptors, gene, descriptor):
+    """Return expressions of a gene with matching descriptor."""
+    indices = get_indices(descriptors, descriptor)
+    return [
+        expression[gene]
+        for i, expression in enumerate(expressions)
+        if i in indices and gene in expression
+    ]
+
+
+def load_expressions(aggregator, expression_fns=[], sep="\t", descriptors=[]):
+    """Read expressions from files."""
+    raw_expressions = [
+        load_expression(expression_fn, sep) for expression_fn in expression_fns
+    ]
+    if aggregator:
+        raw_expressions.extend(aggregator["raw_expressions"])
+        descriptors.extend(aggregator["descriptors"])
+    genes = {key for raw_expression in raw_expressions for key in raw_expression.keys()}
+    grouped_expressions = {
+        gene: {
+            descriptor: get_values(raw_expressions, descriptors, gene, descriptor)
+            for descriptor in sorted(set(descriptors))
+            if get_values(raw_expressions, descriptors, gene, descriptor)
+        }
+        for gene in genes
+    }
+    return raw_expressions, descriptors, grouped_expressions
+
+
+def get_log_expressions(grouped_expressions):
+    """Get log(expression + 1) for all expressions."""
+    log_expressions = {
+        gene: {
+            descriptor: [
+                math.log(expression + 1.0, 2.0)
+                for expression in grouped_expressions[gene][descriptor]
+            ]
+            for descriptor in grouped_expressions[gene]
+        }
+        for gene in grouped_expressions
+    }
+    return log_expressions
+
+
+def generate_statistic(expression, gene, attribute, expression_type):
+    """Get box plot statistic for expressions of a single gene and attribute."""
+    min_val = min(expression)
+    max_val = max(expression)
+    median = np.percentile(expression, 50.0)
+    q1 = np.percentile(expression, 25.0)
+    q3 = np.percentile(expression, 75.0)
+    iqr = q3 - q1
+    lowerwhisker = max(min_val, q1 - 1.5 * iqr)
+    upperwhisker = min(max_val, q3 + 1.5 * iqr)
+    data_count = len(expression)
```

Comment:
Is there any special reason why just not create a dictionary (using expressions above) without first assigning values to variables?